pkg/clientcmd/configutil.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clientcmd
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+)
+
+// This file contains convenience methods for working with the kubeconfig and loading specific configurations from it
+
+// Creates an api.Config by searching for the kubeconfig on the default search path. If an override 'kubeconfigPath' is
+// given then that path is used instead of the default path.
+func NewApiConfigForDefaultSearchPath(kubeconfigPath string) (*api.Config, error) {
+	configLoader := clientcmd.NewDefaultClientConfigLoadingRules()
+	configLoader.ExplicitPath = kubeconfigPath
+	return configLoader.Load()
+}
+
+// Creates an api.Config for a given kubeconfig string.
+func NewApiConfigForKubeconfig(kubeconfig string) (*api.Config, error) {
+	return clientcmd.Load([]byte(kubeconfig))
+}
+
+// Creates a rest.Config by searching for the kubeconfig on the default search path. If an override 'kubeconfigPath' is
+// given then that path is used instead of the default path. If no override is given, an attempt is made to load the
+// 'in cluster' config. If this fails, then the default search path is used.
+func NewRestConfigForDefaultSearchPath(kubeconfigPath string) (*rest.Config, error) {
+	if kubeconfigPath == "" {
+		config, err := rest.InClusterConfig()
+		if err == nil {
+			return config, nil
+		}
+	}
+	apiConfig, err := NewApiConfigForDefaultSearchPath(kubeconfigPath)
```

Comment:
Is it ok to pass in an empty string for the kubeconfig path here? Does the configLoader (above) handle that correctly? Or should we return an error if getting an in-cluster-config above fails?

Reply:
Yes, empty string is expected & handled. If the string is empty then it searches for the kubeconfig in all the normal locations (${KUBECONFIG}, ${HOME}/.kube/config, etc).